util/keydecoder/keydecoder.go

```diff
@@ -0,0 +1,123 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keydecoder
+
+import (
+	"fmt"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/infoschema"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/tablecodec"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// HandleType is the type of a Handle, could be `int`(ie. kv.IntHandle) or `common`(ie. *kv.CommonHandle)
+type HandleType string
+
+// HandleType instances
+const (
+	IntHandle    HandleType = "int"
+	CommonHandle HandleType = "common"
+
+	UnknownHandle HandleType = "unknown"
+)
+
+// DecodedKey is a struct contains detailed information about a key, its json form should be used to fill KEY_INFO field in `DEADLOCKS` and `DATA_LOCK_WAITS`
+type DecodedKey struct {
+	DbID              int64      `json:"db_id,omitempty"`
+	DbName            string     `json:"db_name,omitempty"`
+	TableID           int64      `json:"table_id"`
+	TableName         string     `json:"table_name,omitempty"`
+	HandleType        HandleType `json:"handle_type,omitempty"`
+	IsPartitionHandle bool       `json:"partition_handle,omitempty"`
+	HandleValue       string     `json:"handle_value,omitempty"`
+	IndexID           int64      `json:"index_id,omitempty"`
+	IndexName         string     `json:"index_name,omitempty"`
+	IndexValues       []string   `json:"index_values,omitempty"`
+}
+
+func handleType(handle kv.Handle) HandleType {
+	if _, ok := handle.(kv.IntHandle); ok {
+		return IntHandle
+	} else if _, ok := handle.(*kv.CommonHandle); ok {
+		return CommonHandle
+	} else if h, ok := handle.(kv.PartitionHandle); ok {
+		return handleType(h.Handle)
+	} else if h, ok := handle.(*kv.PartitionHandle); ok {
+		return handleType(h.Handle)
+	} else {
+		logutil.BgLogger().Warn("Unexpected kv.Handle type",
+			zap.Any("handle", handle),
+			zap.String("handle Type", fmt.Sprintf("%T", handle)),
+		)
+	}
+	return UnknownHandle
+}
+
+// DecodeKey decodes `key` (either a Record key or an Index key) into `DecodedKey`, which is used to fill KEY_INFO field in `DEADLOCKS` and `DATA_LOCK_WAITS`
+func DecodeKey(key []byte, infoschema infoschema.InfoSchema) (DecodedKey, error) {
+	var result DecodedKey
+	if !tablecodec.IsRecordKey(key) && !tablecodec.IsIndexKey(key) {
+		return result, errors.Errorf("Unknown key type for key %v", key)
+	}
+	tableID, indexID, isRecordKey, err := tablecodec.DecodeKeyHead(key)
+	if err != nil {
+		return result, err
+	}
+	result.TableID = tableID
+
+	table, tableFound := infoschema.TableByID(tableID)
+
+	// The schema may have changed since when the key is get.
+	// Then we just omit the table name and show the table ID only.
+	// Otherwise, we can show the table name and table ID.
+	if tableFound {
+		result.TableName = table.Meta().Name.O
+
+		schema, ok := infoschema.SchemaByTable(table.Meta())
+		if !ok {
+			return result, errors.Errorf("no schema associated with table which tableID=%d found", tableID)
```

Comment:
Is this returning both error and partial result? Seems the partial result is not added to the SQL queries' result set.

I think it would be good if there can be tests on SQL layer, if possible. At least it should be tested in UTF.

Reply:
> Is this returning both error and partial result? Seems the partial result is not added to the SQL queries' result set because err is not nil.

Now we'll give a warning at once, and return nil for error on those "not so important" errors.

> I think it would be good if there can be tests on SQL layer, if possible. At least it should be tested in UTF.

I prefer a UTF, I always feel (unreasonably) strange when doing test **inside** TiDB with SQL. 😭

Reply:
😥 UTF is fine. Writing them in TiDB repo make those tests run in CI which helps us find problem more in time, but it's often hard to write the code.

Reply:
We could create a UTF test suite add some complex tests or time consuming test cases in it. Actually MySQL does not have unit-tests also but use SQL to do the tests, which is called `mtr test` in MySQL code. 😥